model_compression_toolkit/core/keras/data_util.py

```diff
@@ -29,39 +50,151 @@ def flat_gen_fn(data_gen_fn: Callable[[], Generator]):
     Returns:
         A factory for a flattened data generator.
     """
+
     def gen():
         for inputs_batch in data_gen_fn():
             for sample in zip(*inputs_batch):
-                yield to_tf_tensor(sample)
-    return gen
+                yield tuple([tf.convert_to_tensor(s) for s in sample])
 
+    return gen
 
-# TODO in tf dataset and dataloader are combined within tf.data.Dataset. For advanced use cases such as gptq sla we
-#  need to separate dataset from dataloader similarly to torch data_util.
 class TFDatasetFromGenerator:
```

Comment:
I think trying to build exactly the same objects as in torch is more confusing than helpful. In torch Dataset is indended to be extended, and that's what every custom dataset class does. Tf isn't built the same, and here some "dataset" classes contain tf dataset, and some not, and then for dataloader you convert the dataset back into generator only to then convert back into dataset that serves as dataloader. I suggest we don't try to imitate torch

Reply:
I agree this seems a bit weird, but as you wrote above, "Let's try to be consistent in the logic between frameworks, one way or another." The only reason this is built this way is to make sure the frameworks are aligned, and I do not see a different implementation that can be similar to torch, yet very different from the current implementation (except points you mentioned like FixedSampleInfoDataset and extracting common logic to base class).